### PR TITLE
Remove XP (and Server 2003) support from STL

### LINKED ISSUE #1192: Update OS minimum version and remove XP runtime detection
Currently STL has to target Vista as minimum OS version for x86 and x64

Update `filesystem.cpp`
https://github.com/microsoft/STL/blob/87dc1d3383a052216822fab1a37c8fe2794b0cdb/stl/src/filesystem.cpp#L56-L99

Update awint.hpp / winapisupp.cpp

*Note*: the ABI compatibility for XP toolset should stay. The XP toolset should build; it just doesn't need to support Windows XP anymore.

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Resolves #1192

### CODE REVIEW

File: stl/src/filesystem.cpp [hunk 1]
```diff
@@ -233,28 +125,14 @@ namespace {
...
-                return __std_win_error::_Success;
-            }
-
-            _Last_error = __std_win_error{GetLastError()};
-        }
-#endif // _STL_WIN32_WINNT < _WIN32_WINNT_VISTA
-
-        return _Last_error;
+        return __std_win_error{GetLastError()};
+        ;
```
Comment (StephanTLavavej) [inline stl/src/filesystem.cpp:135]: Spurious semicolon here.

File: stl/src/filesystem.cpp [hunk 2]
```diff
@@ -966,83 +805,43 @@ struct alignas(long long) _Aligned_file_attrs {
...
-
-        if (_Flags != _Attribute_tag_info_data && _Bitmask_includes(_Flags, _Basic_info_data)) {
-            // we have data FileBasicInfo can fill in, that FileAttributeTagInfo wouldn't exactly fill in
-            FILE_BASIC_INFO _Info;
-            if (!_Get_info_fn(_Handle._Get(), FileBasicInfo, &_Info, sizeof(_Info))) {
-                return __std_win_error{GetLastError()};
-            }
+    constexpr auto _Basic_info_data = __std_fs_stats_flags::_Attributes | __std_fs_stats_flags::_Last_write_time;
+    constexpr auto _Attribute_tag_info_data = __std_fs_stats_flags::_Attributes | __std_fs_stats_flags::_Reparse_tag;
+    constexpr auto _Standard_info_data      = __std_fs_stats_flags::_File_size | __std_fs_stats_flags::_Link_count;
```
Comment (AlexGuteniev) [inline stl/src/filesystem.cpp:810]: Weird clang format rules

File: stl/src/winapinls.cpp [hunk 3]
```diff
@@ -1,661 +0,0 @@
...
-        return pfLCMapStringEx(lpLocaleName, dwMapFlags, lpSrcStr, cchSrc, lpDestStr, cchDest, nullptr, nullptr, 0);
-    }
-
-    // ...otherwise fall back to using LCMapString.
-    return LCMapStringW(__crtDownlevelLocaleNameToLCID(lpLocaleName), dwMapFlags, lpSrcStr, cchSrc, lpDestStr, cchDest);
-}
-
-
-// __crtGetLocaleInfoEx() - Wrapper for GetLocaleInfoEx().
-extern "C" int __cdecl __crtGetLocaleInfoEx(
```
Comment (StephanTLavavej) [inline stl/src/winapinls.cpp:652]: It appears that this is changing the export surface, which is a bincompat break:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29115\bin\Hostx86\x86>dumpbin /exports msvcp140.dll | findstr __crtGetLocaleInfoEx
       1501  5DC 0000C1E0 __crtGetLocaleInfoEx
```

This function was declared as `_CRTIMP2`, which exports it. We need to be sure that the export surface of the DLL remains unchanged.

In general, if we're removing any separately compiled functions, we need to be sure that no older user-compiled objects could be looking for those functions (when static linking, or using the import lib).

File: stl/src/filesystem.cpp (hunk 2 above)
Comment (StephanTLavavej) [reply stl/src/filesystem.cpp:810]: We've configured clang-format to align `=` signs and to stay within 120 columns. When forced to choose, clang-format considers `=` alignment to be lower priority.

File: stl/src/winapinls.cpp (hunk 3 above)
Comment (AlexGuteniev) [reply stl/src/winapinls.cpp:652]: I'm seeing this as test failure. Will fix

**Comment by AlexGuteniev** [comment]
Actually probably need to keep not only `_CRTIMP2`, but every function visible from header

**Comment by AlexGuteniev** [comment]
I'll only remove functions in anonymous namespace, others cannot be safely removed.

File: stl/src/awint.hpp [hunk 4]
```diff
@@ -16,17 +16,7 @@ _CRT_BEGIN_C_HEADER
...
-#define __crtFlsAlloc(lpCallback) FlsAlloc(lpCallback)
-
-#define __crtFlsFree(dwFlsIndex) FlsFree(dwFlsIndex)
-
-#define __crtFlsGetValue(dwFlsIndex) FlsGetValue(dwFlsIndex)
-
-#define __crtFlsSetValue(dwFlsIndex, lpFlsData) FlsSetValue(dwFlsIndex, lpFlsData)
-
-#else // _STL_WIN32_WINNT >= _WIN32_WINNT_WS03
+#if _STL_WIN32_WINNT < _WIN32_WINNT_WS03
```
Comment (BillyONeal) [inline stl/src/awint.hpp:19]: Looks like this macro can be completely eliminated now

File: stl/src/awint.hpp [hunk 5]
```diff
@@ -36,88 +26,9 @@ PVOID __cdecl __crtFlsGetValue(__in DWORD dwFlsIndex);
...
-    CompareStringEx(lpLocaleName, dwCmpFlags, lpString1, cchCount1, lpString2, cchCount2, nullptr, nullptr, 0)
-
-#define __crtLCMapStringEx(lpLocaleName, dwMapFlags, lpSrcStr, cchStr, lpDestStr, cchDest) \
-    LCMapStringEx(lpLocaleName, dwMapFlags, lpSrcStr, cchStr, lpDestStr, cchDest, nullptr, nullptr, 0)
-
-#define __crtGetLocaleInfoEx(lpLocaleName, LCType, lpLCData, cchData) \
-    GetLocaleInfoEx(lpLocaleName, LCType, lpLCData, cchData)
-
-#else // _STL_WIN32_WINNT >= _WIN32_WINNT_VISTA
+#if _STL_WIN32_WINNT < _WIN32_WINNT_VISTA
```
Comment (BillyONeal) [inline stl/src/awint.hpp:31]: Ditto no longer necessary.

File: stl/src/primitives.hpp [hunk 6]
```diff
@@ -299,18 +288,16 @@ namespace Concurrency {
...
                 // fall through
             case __stl_sync_api_modes_enum::vista:
-                if (are_vista_sync_apis_available()) {
-                    new (p) stl_critical_section_vista;
-                    return;
-                }
-                // fall through
+                new (p) stl_critical_section_vista;
+                return;
             case __stl_sync_api_modes_enum::concrt:
```
Comment (BillyONeal) [inline stl/src/primitives.hpp:293]: This mode should now be impossible.

File: stl/src/awint.hpp (hunk 4 above)
Comment (AlexGuteniev) [reply stl/src/awint.hpp:19]: Yes! Thought about this a bit more, and understood that only `.cpp` implementation needs to be preserved, not declaration.

**Review by StephanTLavavej — COMMENTED** [review]
This looks good, thanks for working on this extensive set of changes! I'll push fixes for ABI compat and a few other things.

**Comment by StephanTLavavej** [comment]
Awesome - your PR shrinks the STL's main DLL by 1%! This is an extremely rare gift, thank you! :tada:

Changelog for the commits I pushed:
* In `primitives.hpp`, add `static_assert`s for the size/alignment of `stl_MEOW_concrt`, and consolidate the comments into a single TRANSITION comment.
* `winapisupp.cpp` no longer needs `intrin.h`; it was previously needed for `__crtInitOnceExecuteOnce`. (There will be a simple merge conflict with #1270.)
* In `winapisupp.cpp`, `SubmitThreadpoolWork` and `CloseThreadpoolWork` are `void` functions, so we don't need to `return` them.
* Fix ABI: `awint.hpp` was declaring some functions as `_CRTIMP2`, which expands to `__declspec(dllexport)`. To preserve the exported-ness of these functions, I've simply marked their definitions as `_CRTIMP2` (there were a couple of existing occurrences for precedent). This avoids the verbosity of restoring the `awint.hpp` declarations.
* Comment all of the now-unused wrapper functions as "TRANSITION, ABI: preserved for binary compatibility", so they can be immediately eliminated in vNext.
* In `filesystem.cpp`, previously different preprocessor blocks became possible to fuse; I additionally adjusted their style to match the following preprocessor logic.
* In `winapisupp.cpp`, `CreateSymbolicLinkW` was previously being loaded dynamically, which was being changed to a direct call. However, it is [documented](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinkw) as "desktop apps only", and both `filesys.cpp` and `filesystem.cpp` guard their calls with `_CRT_APP`. I'm 98% certain that `winapisupp.cpp` should do the same thing.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I believe I've successfully ported your `stl/CMakeLists.txt` changes to `stl/msbuild` - I pushed a commit.

**Comment by StephanTLavavej** [comment]
Thanks again for this wonderful improvement - ripping out unnecessary complicated code is so awesome. :smile_cat:

**Pull request merged**